### PR TITLE
Arbitrary arity tuple

### LINKED ISSUE #411: Arbitrary long Tuples (*:) produce Quat.Generic instead of Quat.Product
**Version**: 4.8.1
**Module**: `quill-sql`
**Database**: `postgres`

```scala
type MyRow1 = (Int, String)
type MyRow2 = Int *: String *: EmptyTuple

inline def q = quote{
  querySchema[MyRow2]("my_table", t => t._1 -> "int_field", t => t._2 -> "string_field")
}
run(q)
```

### Expected behavior

Both types should work fine.

### Actual behavior

`MyRow1` produces `"SELECT x.int_field, x.string_field FROM my_table"`.
`MyRow1` fails with `io.getquill.quotation.QuatException: Was expecting SQL-level type must be a Product but found '<G>'`.

### Steps to reproduce the behavior

https://scastie.scala-lang.org/i2RFvK3SS7SzIWwJW8fugw

### Workaround

I'm currently using `TupleConverter`:

```scala
type TupleConverter[T <: Tuple] =
  T match
    case EmptyTuple => EmptyTuple
    case a *: EmptyTuple => Tuple1[a]
    case a *: b *: EmptyTuple => (a, b)
    case a *: b *: c *: EmptyTuple => (a, b, c)
    case a *: b *: c *: d *: EmptyTuple => (a, b, c, d)
    case a *: b *: c *: d *: e *: EmptyTuple => (a, b, c, d, e)
    case a *: b *: c *: d *: e *: f *: EmptyTuple => (a, b, c, d, e, f)
```
This helps.

@getquill/maintainers

### PR DESCRIPTION
**Pull request by Primetalk**
Fixes #411

### Problem

To support Scala 3-style tuples, a few changes are needed:
- support at Quat level;
- decode to Scala 3 tuples;
- support pattern matching inside quotes;
- tuple construction (`*:`) inside quotes.

### Solution

So far Quat and decoding are implemented.

There are two variants of decoding:
- using recursive approach;
- using runtime.Tuples.fromIArray.

The latter might be more performant.

### Checklist

- [x] Unit test all changes
- [ ] Update `README.md` if applicable
- [x] Add `[WIP]` to the pull request title if it's work in progress
- [ ] [Squash commits](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one) that aren't meaningful changes

@getquill/maintainers

### CODE REVIEW

File: quill-sql/src/main/scala/io/getquill/generic/GenericDecoder.scala [hunk 1]
```diff
@@ -341,22 +381,50 @@ object ConstructDecoded {
...
-          terms.map(_.asTerm)
-        )
-      // println(s"=========== Create from Tuple Constructor ${Format.Expr(construct.asExprOf[T])} ===========")
-      construct.asExprOf[T]
-      // If we are a case class with no generic parameters, we can easily construct it
+    if (tpe <:< TypeRepr.of[EmptyTuple]) {
+      '{EmptyTuple}
+    } else if (tpe <:< TypeRepr.of[Tuple]) {
+      '{scala.runtime.Tuples.fromIArray(IArray(${Varargs(terms)})).asInstanceOf[T]}
+      // Alternative version:
```
Comment (guizmaii) [inline quill-sql/src/main/scala/io/getquill/generic/GenericDecoder.scala:388]: Why keeping commented code? 🤔

Comment (Primetalk) [reply quill-sql/src/main/scala/io/getquill/generic/GenericDecoder.scala:388]: I wanted to consult which version is preferred - array based or recursion-based.

File: quill-sql/src/main/scala/io/getquill/parser/Parser.scala [hunk 2]
```diff
@@ -153,6 +155,126 @@ class ValParser(val rootParse: Parser)(using Quotes, TranspileConfig)
...
+ * }
+ * }}}
+ */
+object ArbitraryTupleConstructionInlined {
+  def unapply(using Quotes)(t: quotes.reflect.Term): Option[(quotes.reflect.Term, quotes.reflect.Term)] = {
+    import quotes.reflect.{Ident => TIdent, *}
+    t match {
+      case
+        Inlined(
+        //  Some(Apply(
```
Comment (Primetalk) [inline quill-sql/src/main/scala/io/getquill/parser/Parser.scala:202]: Question: should we also match on `call` part or is it enough to match just the body?

**Comment by deusaquilus** [comment]
This is really amazing work! Thank you!

**Pull request merged**